lib/ulid/generate.rb

```diff
@@ -1,84 +1,35 @@
+# frozen_string_literal: true
+
 require 'securerandom'
 require 'ulid/constants'
 
 module ULID
-
   module Generate
     include Constants
 
     # returns the binary ULID as Base32 encoded string.
     def encode32
-      # Optimized unrolled loop ahead.
-      # From https://github.com/RobThree/NUlid
-      # via https://github.com/oklog/ulid/blob/c3c01856f7e43fa64133819126887124d5f05e39/ulid.go#L154
-
-      id = @bytes.bytes
-
-      output = +''
-
-      # Base32 encodes 5 bits of each byte of the input in each byte of the
-      # output. That means each input byte produces >1 output byte and some
-      # output bytes encode parts of multiple input bytes.
-      #
-      # The end result is a human / URL friendly output string that can be
-      # decoded into the original bytes.
+      (hi, lo) = bytes.unpack("Q>Q>")
+      value = (hi << 64) | lo
 
-      # 10 byte timestamp
-      output << ENCODING[(id[0]&224)>>5]
-      output << ENCODING[id[0]&31]
-      output << ENCODING[(id[1]&248)>>3]
-      output << ENCODING[((id[1]&7)<<2)|((id[2]&192)>>6)]
-      output << ENCODING[(id[2]&62)>>1]
-      output << ENCODING[((id[2]&1)<<4)|((id[3]&240)>>4)]
-      output << ENCODING[((id[3]&15)<<1)|((id[4]&128)>>7)]
-      output << ENCODING[(id[4]&124)>>2]
-      output << ENCODING[((id[4]&3)<<3)|((id[5]&224)>>5)]
-      output << ENCODING[id[5]&31]
+      # use the RFC4648 Base32 encoding and convert to Crockford's Base32 with a simple translate
+      # assumes that the value is a 128-bit integer
+      # also assumes that .to_s(32) is lowercase
+      b32 = value.to_s(32).tr!(B32_RCF4648_FRAGMENT, B32_CROCKFORD_FRAGMENT).upcase!
```

Comment:
```suggestion
      b32 = value.to_s(32)
      b32.tr!(B32_RCF4648_FRAGMENT, B32_CROCKFORD_FRAGMENT)
      b32.upcase!
```
you can't chain `!` methods because most of them return `nil` if no change happened.

Reply:
it just happens that these ! methods are chainable builder style methods. :shrug:

Reply:
Alright, I would have told you so 🤷

Reply:
@colinbendell if the `tr!` call doesn't match anything, it will return `nil` and the `upcase!` call will fail. That's the pitfall to watch out for.